web/components/FcPaneMap.vue

```diff
@@ -234,30 +234,87 @@ export default {
+      const waypoints = this.locationsSelectionForMode.locations;
+
+      let waypointIndex = 0;
+      let waypoint = waypoints[waypointIndex];
+      const features = [];
+      this.locationsForMode.forEach((location) => {
         const {
           geom,
           lat,
           lng,
-          ...properties
+          ...propertiesRest
         } = location;
+
         const geometry = {
           type: 'Point',
           coordinates: [lng, lat],
         };
-        const { multi } = this.locationMode;
-        properties.multi = multi;
-        if (multi) {
-          properties.locationIndex = i;
-          properties.selected = i === this.locationEditIndex;
+
+        /*
+         * It is possible for the current location to match multiple consecutive waypoints.
+         * We could forbid selecting the same location multiple times, but that would introduce
+         * a lot of validation complexity in both frontend and backend.  It would also make it
+         * impossible to select a corridor that loops back on itself.
+         */
+        const waypointIndices = [];
+        while (waypointIndex < waypoints.length
+          && propertiesRest.centrelineType === waypoint.centrelineType
+          && propertiesRest.centrelineId === waypoint.centrelineId) {
+          waypointIndices.push(waypointIndex);
+          waypointIndex += 1;
+          waypoint = waypoints[waypointIndex];
         }
-        return { type: 'Feature', geometry, properties };
+
+        /*
+         * `locationIndex === null` here indicates that the location is not a waypoint, and
+         * can be drawn using a corridor marker.
+         */
+        let locationIndex = -1;
+        let selected = false;
+        const k = waypointIndices.length;
+        if (k === 0) {
+          if (propertiesRest.centrelineType === CentrelineType.SEGMENT) {
+            // We only show corridor markers at intersections.
+            return;
+          }
+        } else if (waypointIndices.includes(this.locationsEditIndex)) {
+          /*
+           * In this case, we might have several consecutive waypoints at the same location,
+           * which might cause the currently selected waypoint to be hidden - so we prioritize
+           * it.
+           */
+          locationIndex = this.locationsEditIndex;
+          selected = true;
+        } else {
+          /*
+           * Here there is no selected waypoint, so we just show the last matching waypoint.
+           */
+          locationIndex = waypointIndices[k - 1];
+        }
+
+        const { multi } = this.locationMode;
+        const properties = {
+          locationIndex,
+          multi,
+          selected,
+          ...propertiesRest,
+        };
+        const feature = { type: 'Feature', geometry, properties };
+        features.push(feature);
       });
+
       return {
         type: 'FeatureCollection',
         features,
       };
     },
+    locationsSelectionForMode() {
```

Comment:
can we make this and `locationsForMode` getters as well?  after all, they depend exclusively on `vuex` state, and are likely to be used elsewhere once we start to build out View Data.